### PR TITLE
Update engineer full access role

### PR DESCRIPTION
**Pull request by cat5inthecradle**
* Align `Engineer_FullAccess` role more closely with the older `Developer` role.
* Tune permissions to be more least-privilege

Specifically:
1. Engineering_FullAccess
   - Allowing PassRole, to allow CloudFormation deploys
2. Engineering_FullAccess and Engineering_ReadOnly
   - Removing the ability to delete production/test secrets

## Testing story

```
> bundle exec rake stack:iam:validate RAILS_ENV=production
...
Pending update for stack `IAM`:
Add CloudformationPassRolePolicy [AWS::IAM::ManagedPolicy] 
Modify ContributorS3Policy [AWS::IAM::ManagedPolicy] Properties (PolicyDocument)
Remove DenySecretsPolicy [AWS::IAM::ManagedPolicy] 
Add EngineerSecretsPolicy [AWS::IAM::ManagedPolicy] 
Modify EngineeringFullAccessRole [AWS::IAM::Role] Properties (ManagedPolicyArns, ManagedPolicyArns, ManagedPolicyArns)
Modify EngineeringReadOnlyRole [AWS::IAM::Role] Properties (ManagedPolicyArns, ManagedPolicyArns)
Finished stack:iam:validate (1 minute)
```

## Deployment strategy

Before merging PR, pull from staging and deploy per README.md instructions. If successful, merge PR. If not, troubleshoot and retry.

## Follow-up work

Anticipate future issue where Engineers can't read `CfnStack/adhoc-*` secrets. They never had permission permission, but I removed the ineffective allow that implied that they should.

## PR Checklist:

- [ ] ~Tests provide adequate coverage~
- [x] Privacy and Security impacts have been assessed
- [x] Code is well-commented
- [ ] ~New features are translatable or updates will not break translations~
- [x] Relevant documentation has been added or updated
- [x] User impact is well-understood and desirable
- [x] Pull Request is labeled appropriately
- [x] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: aws/cloudformation/iam.yml.erb [hunk 1]
```diff
@@ -306,7 +306,7 @@ roles = {
     GoogleIdsSSMParameter: 'DeveloperGoogleIDs',
     ManagedPolicyArns: [
       'arn:aws:iam::aws:policy/ReadOnlyAccess',
-      '!Ref DenySecretsPolicy'
+      '!Ref EngineerSecretsPolicy',
```
Comment (cat5inthecradle) [inline aws/cloudformation/iam.yml.erb]: This is a rename, and will be deployed as follows by Cloudformation:

- create new policy with new name
- update role to use new policy
- delete old, now unused, policy

File: aws/cloudformation/iam.yml.erb [hunk 2]
```diff
@@ -583,7 +613,6 @@ roles.each do |name, config| %>
             Resource:
               - !Sub arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:development/cdo/*
               - !Sub arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:adhoc/cdo/*
-              - !Sub arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:CfnStack/adhoc-*
```
Comment (cat5inthecradle) [inline aws/cloudformation/iam.yml.erb]: This line was useless, overruled by the `Deny` on line 580

File: aws/cloudformation/iam.yml.erb [hunk 3]
```diff
@@ -563,16 +591,18 @@ roles.each do |name, config| %>
...
+      ManagedPolicyName: EngineerSecretsPolicy
       Path: /engineering/
       PolicyDocument:
         Version: 2012-10-17
         Statement:
           - Effect: Deny
-            Action: secretsmanager:GetSecretValue
+            Action:
+              - secretsmanager:GetSecretValue
+              - secretsmanager:DeleteSecret
```
Comment (sureshc) [inline aws/cloudformation/iam.yml.erb]: Should we add the Stack-specific Secrets as well? They have the Secret ID/Name `CfnStack/[stack name]/*`

File: aws/cloudformation/iam.yml.erb [hunk 4]
```diff
@@ -390,6 +392,32 @@ roles.each do |name, config| %>
...
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Action: iam:PassRole
+            Resource: '*'
+            Condition:
+              StringEquals:
+                iam:PassedToService: cloudformation.amazonaws.com
+          # Require CloudFormation Service Role on all stack operations.
```
Comment (unlox775-code-dot-org) [inline aws/cloudformation/iam.yml.erb]: What is the effect of this deny rule? Ultimately to me it looks like we're denying create, update, and delete and create change set for everything except a certain transform.

I'm not quite sure what the string not equals is doing. Overall it just doesn't seem to add up to what the description says of requiring it on all stack operations.  Or my brain is twisted up like it does, with negative-rule logic. Can we update the comment to maybe explain this better?

File: aws/cloudformation/iam.yml.erb (hunk 3 above)
Comment (unlox775-code-dot-org) [reply aws/cloudformation/iam.yml.erb]: Isn't that already included?

```
              - !Sub arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:CfnStack/*
```

File: aws/cloudformation/iam.yml.erb (hunk 4 above)
Comment (cat5inthecradle) [reply aws/cloudformation/iam.yml.erb]: It denies cloudformation stack modification unless you tell Cloudformation to assume a certain role for it's stack modifications. Normally you'll just run these operations with your own role. I think the CloudFormationService role is _more_ powerful than the engineer's role, and it gives us a place to grant permissions useful for CI/CD that we don't necessarily want to grant to everyone.

- but it lets you work with serverless transforms (not 100% sure the scenario here, but surely related to Lambda custom resources, or perhaps Javabuidler)

It's an IAM setup that predates me, so I can't fully speak to the original intent.

File: aws/cloudformation/iam.yml.erb (hunk 3 above)
Comment (cat5inthecradle) [reply aws/cloudformation/iam.yml.erb]: Yes, this is denying Get and Delete for `CfnStack/*`

**Review by unlox775-code-dot-org — CHANGES_REQUESTED** [review]
What if we were to re-write this rule, to avoid negative "block list" style usage, like the `Deny` and `StringNotEquals`.  I also may be misunderstanding (let me know), but it seems to me that the intent is to only allow this one role to pass to 
CloudFormationService, and to only call one of: CreateStack, UpdateStack, DeleteStack, CreateChangeSet.

What if we were to change this role policy to this:
```yaml
  CloudformationPassRolePolicy:
    Type: AWS::IAM::ManagedPolicy
    Properties:
      ManagedPolicyName: CloudformationPassRolePolicy
      Path: /admin/
      PolicyDocument:
        Version: 2012-10-17
        Statement:
          - Effect: Allow
            Action: iam:PassRole
            Resource: 'arn:aws:iam::${AWS::AccountId}:role/admin/CloudFormationService'
            Condition:
              StringEquals:
                iam:PassedToService: cloudformation.amazonaws.com
          # Require CloudFormation Service Role on all stack operations.
          - Effect: Allow
            Action:
              - cloudformation:CreateStack
              - cloudformation:UpdateStack
              - cloudformation:DeleteStack
              - cloudformation:CreateChangeSet
            Resource: "arn:aws:cloudformation:*:aws:transform/Serverless-2016-10-31"
            Condition:
              StringEquals:
                cloudformation:RoleARN: !Sub "arn:aws:iam::${AWS::AccountId}:role/admin/CloudFormationService"
```

Overall I would like to consider moving away from negative-logic block-list style.  For example in this, the PR's code reads to the me that we are Deny'ing unless you are doing you are in the safe zone of:
- role/admin/CloudFormationService
- AND transform/Serverless-2016-10-31

But really it is saying (because it is really an AND):
- I can do CreateStack, etc to any resource, as long as I don't use role/admin/CloudFormationService
- AND I can do CreateStack, etc. with any role I want as long as I am using transform/Serverless-2016-10-31 as my resource.

And the deny logic in general doesn't stop me at all from calling other CloudFormation methods like CreateStackSet, etc.  I ran a POC of this here: https://github.com/unlox775/security_poc/blob/main/passrole_star/

**Comment by cat5inthecradle** [comment]
We're building on top of "[PowerUserAccess](https://us-east-1.console.aws.amazon.com/iamv2/home?region=us-east-1#/policies/details/arn%3Aaws%3Aiam%3A%3Aaws%3Apolicy%2FPowerUserAccess?section=permissions)" which Allows these cloudformation 'write' actions on all resources. Simply switching this to an Allow-based policy would break the protection this is intending to provide.

I hopped into the visual editor, and I think we could perhaps get rid of the `NotResource` part. Here's a policy which denies all Write actions with wildcard resource targets that should not match transforms.

Here's the [Slack thread](https://codedotorg.slack.com/archives/C03CK49G9/p1628799470158400) and [pull request](#42044 ) with the context behind adding the `NotResource`.

```yaml
Version: "2012-10-17"
Statement:
  - Effect: "Deny"
    Action:
      - "cloudformation:CreateUploadBucket"
      - "cloudformation:RegisterType"
      - "cloudformation:SetTypeDefaultVersion"
      - "cloudformation:RegisterPublisher"
      - "cloudformation:ActivateType"
      - "cloudformation:DeactivateOrganizationsAccess"
      - "cloudformation:PublishType"
      - "cloudformation:DeactivateType"
      - "cloudformation:SetTypeConfiguration"
      - "cloudformation:DeregisterType"
      - "cloudformation:ActivateOrganizationsAccess"
      - "cloudformation:TestType"
      - "cloudformation:CreateStackSet"
    Resource: "*"
    Condition:
      StringNotEquals:
        cloudformation:RoleARN: !Sub "arn:aws:iam::${AWS::AccountId}:role/admin/CloudFormationService"
  - Effect: "Deny"
    Action:
      - "cloudformation:SetStackPolicy"
      - "cloudformation:ImportStacksToStackSet"
      - "cloudformation:DeleteStackInstances"
      - "cloudformation:CancelUpdateStack"
      - "cloudformation:UpdateStackInstances"
      - "cloudformation:SignalResource"
      - "cloudformation:UpdateTerminationProtection"
      - "cloudformation:RecordHandlerProgress"
      - "cloudformation:UpdateStackSet"
      - "cloudformation:CreateChangeSet"
      - "cloudformation:DeleteStackSet"
      - "cloudformation:CreateStackInstances"
      - "cloudformation:DeleteChangeSet"
      - "cloudformation:ContinueUpdateRollback"
      - "cloudformation:RollbackStack"
      - "cloudformation:CreateStack"
      - "cloudformation:DeleteStack"
      - "cloudformation:UpdateStack"
      - "cloudformation:StopStackSetOperation"
      - "cloudformation:ExecuteChangeSet"
    Resource: 
      - "arn:aws:cloudformation:*:${AWS::AccountId}:stackset-target/*"
      - "arn:aws:cloudformation:*:${AWS::AccountId}:stackset/*:*"
      - "arn:aws:cloudformation:*:${AWS::AccountId}:type/resource/*"
      - "arn:aws:cloudformation:*:${AWS::AccountId}:stack/*/*"
    Condition:
      StringNotEquals:
        cloudformation:RoleARN: !Sub "arn:aws:iam::${AWS::AccountId}:role/admin/CloudFormationService"
  - Effect: "Deny"
    Action:
      - "cloudformation:CreateUploadBucket"
      - "cloudformation:RegisterType"
      - "cloudformation:SetTypeDefaultVersion"
      - "cloudformation:RegisterPublisher"
      - "cloudformation:ActivateType"
      - "cloudformation:DeactivateOrganizationsAccess"
      - "cloudformation:PublishType"
      - "cloudformation:DeactivateType"
      - "cloudformation:SetTypeConfiguration"
      - "cloudformation:DeregisterType"
      - "cloudformation:ActivateOrganizationsAccess"
      - "cloudformation:TestType"
      - "cloudformation:CreateStackSet"
    Resource: "*"
    Condition:
      StringNotEquals:
        cloudformation:RoleARN: !Sub "arn:aws:iam::${AWS::AccountId}:role/admin/CloudFormationService"
  - Effect: "Deny"
    Action:
      - "cloudformation:SetStackPolicy"
      - "cloudformation:ImportStacksToStackSet"
      - "cloudformation:DeleteStackInstances"
      - "cloudformation:CancelUpdateStack"
      - "cloudformation:UpdateStackInstances"
      - "cloudformation:SignalResource"
      - "cloudformation:UpdateTerminationProtection"
      - "cloudformation:RecordHandlerProgress"
      - "cloudformation:UpdateStackSet"
      - "cloudformation:CreateChangeSet"
      - "cloudformation:DeleteStackSet"
      - "cloudformation:CreateStackInstances"
      - "cloudformation:DeleteChangeSet"
      - "cloudformation:ContinueUpdateRollback"
      - "cloudformation:RollbackStack"
      - "cloudformation:CreateStack"
      - "cloudformation:DeleteStack"
      - "cloudformation:UpdateStack"
      - "cloudformation:StopStackSetOperation"
      - "cloudformation:ExecuteChangeSet"
    Resource: 
      - "arn:aws:cloudformation:*:${AWS::AccountId}:stackset-target/*"
      - "arn:aws:cloudformation:*:${AWS::AccountId}:stackset/*:*"
      - "arn:aws:cloudformation:*:${AWS::AccountId}:type/resource/*"
      - "arn:aws:cloudformation:*:${AWS::AccountId}:stack/*/*"
    Condition:
      StringNotEquals:
        cloudformation:RoleARN: !Sub "arn:aws:iam::${AWS::AccountId}:role/admin/CloudFormationService"

```

**Pull request closed without merging**